### ISSUE TITLE
netconfd returns "invalid protocol framing characters received" for get-config request

### ISSUE THREAD

**Issue by anukulverma** [OpenClovis/OpenYuma#28]
Hi,

I m trying to use openssh commands to get device configuration.

hello message sent :
<?xml version="1.0" encoding="UTF-8"?>
<hello xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
  <capabilities>
    <capability>
      urn:ietf:params:netconf:base:1.1
    </capability>
  </capabilities>
</hello>
]]>]]>

get-config payload:
#239
<?xml version="1.0" encoding="UTF-8"?>
<rpc message-id="2"
  xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
  <get-config xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
    <source>
      <running/>
    </source>
  </get-config>
</rpc>
##

Netconfd server (Openyuma) returns following error:
#581 
<?xml version="1.0" encoding="UTF-8"?> 
<rpc-reply xmlns:ncx="http://netconfcentral.org/ns/yuma-ncx" 
  xmlns="urn:ietf:params:xml:ns:netconf:base:1.0"> 
  <rpc-error> 
    <error-type>transport</error-type> 
    <error-tag>malformed-message</error-tag> 
    <error-severity>error</error-severity> 
    <error-app-tag>data-invalid</error-app-tag> 
    <error-message xml:lang="en">invalid protocol framing characters received</error-message> 
    <error-info> 
      <error-number xmlns="http://netconfcentral.org/ns/yuma-ncx">378</error-number> 
    </error-info> 
  </rpc-error> 
</rpc-reply> 
##

**Comment by hoangle** [comment]
1. Please make sure netconf-subsytem running over SSH and sshd is listening to port 830:
netstat -tulpn | grep :830

2. Switch to use ncclient to get device running config
$pip install ncclient or download from https://github.com/ncclient/ncclient

**Comment by anukulverma** [comment]
1) With base 1.0 everything works.
  <?xml version="1.0" encoding="UTF-8"?>
<hello xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
  <capabilities>
    <capability>
      urn:ietf:params:netconf:base:1.0
    </capability>
  </capabilities>
</hello>
]]>]]>

<?xml version="1.0" encoding="UTF-8"?>
<rpc message-id="2"
  xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
  <get-config xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
    <source>
      <running/>
    </source>
  </get-config>
</rpc>
]]>]]>

So , not issue with connection or port.

2. I m using yangcli, ncclient, netopeer -> these r working fine for me. I m trying to develop my own client app for that i m trying with ssh commands.

**Comment by hoangle** [comment]
Yes, netconf server (OpenYuma) is currently support base:1.0 and cause conflict with your client base:1.1. 
So, you should get capabilities from server and switch to use appropriate version as same as ncclient.

**Comment by anukulverma** [comment]
Netconf server is sending both 1.1 & 1.0 to client in hello message,

<?xml version="1.0" encoding="UTF-8"?>
<hello xmlns:nc="urn:ietf:params:xml:ns:netconf:base:1.0"
  xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
  <capabilities>
    <capability>urn:ietf:params:netconf:base:1.0</capability>
    <capability>urn:ietf:params:netconf:base:1.1</capability>

**Comment by hoangle** [comment]
Please turn on debug log (i.e debug4) at netconf server to see a root cause

**Comment by anukulverma** [comment]
Server debug log:

ses_accept_input on session 5
ses read OK (173) on session 5
ses: accept base:1.1 buffer (173):
#166
<?xml version="1.0" encoding="UTF-8"?><rpc message-id="199" xmlns="urn:ietf:params:xml:ns:netconf:base:1.0"><get-config><source><running/></source></get-config></rpc>


ses: invalid base:1;1 framing (inbetween: expect newline)
agt_ncxserver: input failed for session 5 (invalid protocol framing characters received)
ses_msg send 1.1 buff:591
lity
#581
<?xml version="1.0" encoding="UTF-8"?>
<rpc-reply xmlns:ncx="http://netconfcentral.org/ns/yuma-ncx"
  xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
  <rpc-error>
    <error-type>transport</error-type>
    <error-tag>malformed-message</error-tag>
    <error-severity>error</error-severity>
    <error-app-tag>data-invalid</error-app-tag>
    <error-message xml:lang="en">invalid protocol framing characters received</error-message>
    <error-info>
      <error-number xmlns="http://netconfcentral.org/ns/yuma-ncx">378</error-number>
    </error-info>
  </rpc-error>
</rpc-reply>
agt_sys: generating <sysSessionEnd> notification
Queing <sysSessionEnd> notification to send (id: 11)
Event Payload:
  sys:userName root
  sys:sessionId 5
  sys:remoteHost 10.12.23.28
  sys:terminationReason other
Session 5 closed

**Comment by hoangle** [comment]
> ses: invalid base:1;1 framing (inbetween: expect newline)
> agt_ncxserver: input failed for session 5 (invalid protocol framing characters received)
> ses_msg send 1.1 buff:591
Does LF encode as '\n' (wiki: https://tools.ietf.org/html/rfc6242#section-4.2)

**Comment by anukulverma** [comment]
I saw this and copied payload from RFC only. Can u please send me the payload, if i m doing something wrong, i will compare and correct.

**Comment by hoangle** [comment]
I think you have to follow an payload format:
\n#<msgID>\n
...\n
...\n
...\n
\n##\n

**Comment by anukulverma** [comment]
I have tried this also.

agt_hello: set protocol to base:1.1
Session 11 for root@10.12.23.28 now active (base:1.1)
ses_accept_input on session 11
ses read OK (132) on session 11
ses: accept base:1.1 buffer (132):
\n#4\n
<rpc
\n#18\n
 message-id="102"\n
\n#79\n
     xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">\n
  <close-session/>\n
</rpc>


ses: invalid base:1;1 framing (inbetween: expect newline)
agt_ncxserver: input failed for session 11 (invalid protocol framing characters received)
ses_msg send 1.1 buff:591
lity
#581
<?xml version="1.0" encoding="UTF-8"?>
<rpc-reply xmlns:ncx="http://netconfcentral.org/ns/yuma-ncx"
  xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
  <rpc-error>
    <error-type>transport</error-type>
    <error-tag>malformed-message</error-tag>
    <error-severity>error</error-severity>
    <error-app-tag>data-invalid</error-app-tag>
    <error-message xml:lang="en">invalid protocol framing characters received</error-message>
    <error-info>
      <error-number xmlns="http://netconfcentral.org/ns/yuma-ncx">378</error-number>
    </error-info>
  </rpc-error>
</rpc-reply>
agt_sys: generating <sysSessionEnd> notification
Queing <sysSessionEnd> notification to send (id: 23)
Event Payload:
  sys:userName root
  sys:sessionId 11
  sys:remoteHost 10.12.23.28
  sys:terminationReason other
Session 11 closed

**Comment by anukulverma** [comment]
Is it possible for you to help me, to get it working?

**Comment by hoangle** [comment]
Sorry, what am I suggest is to read carefully at https://tools.ietf.org/html/rfc6242#section-4.2 since I never try this before.
Try your best!

**Comment by hoangle** [comment]
Try this:
`
\n#4\n
<rpc
\n#18\n
 message-id="102"\n
\n#79\n
     xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">\n
  <close-session/>\n
</rpc>
\n##\n
`

**Comment by anukulverma** [comment]
same error

**Comment by vlutsenko** [comment]
Hello @anukulverma, I understand that issue happened while ago, but would much appreciate if can you share the root cause and how to overcome.
